### DISCUSSION TITLE
Can't figure out how to properly discard the alpha channel

### DISCUSSION THREAD

**Discussion by Nakilon** [libvips/ruby-vips#355] [Q&A] [answered]
I again appear to be dumber that I suppose when I deal with images and my laziness or lameness plays with me. So I still can't figure out the alpha channel thing.

I want to just properly convert 4 channels to 3 channels.

The issue is... see these two images:
* https://drive.google.com/file/d/1SeYIbaawN5zaR4K3gsoHZyjcb-oy5gri/view?usp=sharing
* https://drive.google.com/file/d/1AMa-1Q8hlfwevHSM1nXHSLZ6l4Dc_rth/view?usp=sharing

If I do
```ruby
image = image.has_alpha? ? image.flatten : image
```
one image becomes just black.

If I do:
```ruby
image = (image.has_alpha? ? image.flatten.composite2(image.bandsplit[3], :screen) : image)
```
another one becomes just white.

I guess I totally misunderstand the thing.

UPD: Note: one image is downloaded from tg stickers, another one is a html5 base64 exported from p5js canvas after using the text and geometry primitives. I.e. none is made by me band by band so I suppose the images aren't "corrupted" in any way.

**Comment by jcupitt** [comment, marked as answer]
Hey @Nakilon,

Your first image has RGB all zero, and only has non zero values in the alpha channel. I find [vipsdisp](https://flathub.org/apps/details/org.libvips.vipsdisp) useful for looking at images like this: it will show you exactly what's in every pixel in the image:

![image](https://user-images.githubusercontent.com/580843/209337495-3c1df3ea-58ae-460d-908b-3a3be630a1f7.png)

By default, `vips_flatten()` will flatten against black (ie. 0), so the output is a black image on a black background :( If you flatten against 255 instead, you'll be able to see the alpha:

```ruby
image = image.flatten(background: 255) if image.has_alpha?
```

That should work for your second image too.

Maybe the black default for flatten wasn't a great idea :(

**Reply by Nakilon** [reply]
Woah, 4 channel eyedropper! Is it available on macOS though?

Thank you for the prompt response and making me see the problem.

I believe I saw the mention of the "default background" somewhere in https://en.wikipedia.org/wiki/Alpha_compositing when I was solving the issue with that image few weeks ago, can't find it now.
That is interesting because on canvas the default color is white I guess (while as you showed it stores the color as black at least in the export by `document.getElementsByTagName('canvas')[0].toDataURL()`). Also it's white in macOS preview. It's white in Telegram. But black in Skype (when you click/zoom it).

**Reply by jcupitt** [reply]
You'd need to compile from source I guess :( Though I've not tried.

https://github.com/jcupitt/vipsdisp
